### DISCUSSION TITLE
Should -BeIn loop with HashTable keys.

### DISCUSSION THREAD

**Discussion by ghost** [pester/Pester#2263] [General]
Hi, I've been using Pester (v5.3.3) for a few weeks now and I'm encountering a strange behavior.

When I try to check a key in a hashtable, the command returns nothing and the process grows in memory.

I found a workaround by using a clone, which is strange because the negative assertion -Not -BeIn works fine.

You can test it:
```
Describe 'Test hashtable keys' {

  BeforeAll {
    $ht = @{
      Test1 = 'OK'
      Test2 = 'OK'
    }
  }

  It 'Test1 pass' {
    'Test1' | Should -BeIn $ht.Keys
  }

  It 'Test3 -not pass' {
    'Test3' | Should -Not -BeIn $ht.Keys
  }

  It 'Test3 clone pass' {
    'Test3' | Should -BeIn $ht.Keys.Clone()
  }

  It 'Test3 infinite loop' {
    'Test3' | Should -BeIn $ht.Keys
  }

}
```

**Comment by fflaten** [comment]
Thanks for the great report. I'm able to reproduce it. 🙂 

A helper-function `Format-Collection` tries to use index-based looping through the `KeyCollection` which doesn't implement `IList` (index-support). It ends up returning a `KeyCollection` on index zero which then end up in the same `Format-Collection` again +++

Easily explained by this sample:
```ps1
(@{ "foo" = $true; "bar" = $false }).Keys[0][0][0][0]
foo
bar
```

`Clone()` is just a shallow clone and apparently returns an array
```ps1
(@{ "foo" = $true; "bar" = $false }).Keys.Clone().Gettype()
IsPublic IsSerial Name                                     BaseType
-------- -------- ----                                     --------
True     True     Object[]                                 System.Array

(@{ "foo" = $true; "bar" = $false }).Keys.Clone()[0]
foo
```

**Reply by fflaten** [reply]
Fix submitted in PR #2265
